core/src/main/java/org/elasticsearch/search/internal/ShardSearchLocalRequest.java

```diff
@@ -224,4 +225,13 @@ public BytesReference cacheKey() throws IOException {
         // we could potentially keep it without copying, but then pay the price of extra unused bytes up to a page
         return out.bytes().copyBytesArray();
     }
+
+    @Override
+    public void rewrite(QueryShardContext context) throws IOException {
+        SearchSourceBuilder rewritten =null;
+        while (rewritten != source) {
```

Comment:
hmm can we make source a local var and only assign it once we are successfully rewritten and exited the loop